### PR TITLE
Make it possible for project contributors to have affiliations on multiple levels and several affiliations

### PR DESCRIPTION
**Pull request by denektenina**
# Description

Link to Jira issue: [https://sikt.atlassian.net/browse/NP-47086](https://sikt.atlassian.net/browse/NP-47086)

This is a fix of one of the missing requirements discovered by testers, where each project participant should be able to have several affiliations on several levels. In the current version it is only possible to have one affiliation and on one level. 

_SORRY for the big PR! All of this had to be delivered together unless we wanted to introduce a another beta-tag (since we also have another version of project in beta :D)_

_Feel free to reach out if you want me to demonstrate the stuff below because you don't want to read it!_

Before this PR:

image

Now:

image

**New functionality / things to test**
* You can edit an affiliation and set it to any level
* You can add multiple affiliations
* You can delete existing affiliation, as long as it's not your last one (you must always have one)
* You can still add a new contributor. If you do, you have to explicitly set name and affiliation for them. If you don't do that, there will be an error next to the field once the field is touched (i.e. when you go to the next tab)
* You can still delete a contributor
* Pagination on contributors
* This works in all views (create new project from my page, create new project from registration view, edit project from project list)

**Suggested order for reviewing:**
1) `ProjectFormDialog`. This is the modal that opens when you want to create or edit a project.
2) `ProjectFormPanel1` - The first tab in the modal - added new component for the new contributor view and removed the old one. Parts of the old one is moved to `ProjectContributors`.
3) `ProjectContributors` - Uses parts from `ProjectFormPanel1` and added logic for pagination and listing of contributors. Button for adding contributor has changed look and is moved to the top. The contributors are now in a paginated table.
4) `ContributorRow` - also a new component but uses project- and name-field that previously was in the now deleted component `ProjectContributorRow ` (you can have a look at that one now as well)
5) `ProjectOrganizationBox` - the white boxes with information about an organization. This is copied from the `OrganizationBox` used in the registration view. The reason I copied it was because it needed some modifications (texts and which EditAffiliationModal it used mainly), and I didn't want to introduce too many conditions and risk intoducing bugs for all the others using the component. Besides, the component mainly consists of other components, so it wasn't too much duplicate code, only duplicate structure. Also, changes in one does not automatically mean the other needs the same changes, so they should live their separate lives.
6) `ProjectEditAffiliationModal`: this is a copy of the `EditAffiliationModal` for the same reasons as above. The differences is the texts, that the ProjectEditAffiliationModal works against a CristinProject-form while `EditAffiliationModal` works against a Registration form, which means it would have to be made generic in a way that could introduce bugs. They also had different functions for editAffiliation, so they would have had to be pulled out, and they were the biggest part of the component. Also, changes in one does not automatically mean the other needs the same changes, so they should live their separate lives. The two components have little rendering logic, and uses the same components. To make the code duplication smaller i extracted parts of the view into an `AuthorName` component and used it by both `ProjectEditAffiliationModal` and `EditAffiliationModal`.
7) `SelectInstitutionForm` is the central component in both components above and existed before this PR. The important change here is that I have to introduce a variable for storing the labels of an organization so that I could have that information available when saving a project contributor affiliation. Thus, `saveAffiliation` can now optionally accept a LanguageString which the receiving function can use. Functions that don't need it are not affected since it's optional.
8) `OrganizationAccordion` - also existed before this PR and received some small changes to make it possible to send labels. 
9) `ProjectAddAffiliationModal` - this is the modal that opens when you want to add another affiliation. This is a copy of `AddAffiliationModal` for the same reasons as for `ProjectEditAffiliationModal`. 

# Checklist

Ensure that the changes are aligned with the expectations of PO/designer before marking the PR as ready for review.

Also ensure that the following criterias are met:

- [x] The changes are working as expected
- [x] The changes are tested OK for different screen sizes
- [x] The changes are tested OK for a11y
- [x] Interactive elements have data-testids
- [x] I have done a QA of my own changes

_Note: some of these criterias may not always be relevant, and can simply be marked as completed._

### CODE REVIEW

File: src/components/AuthorName.tsx
Comment (denektenina) [inline src/components/AuthorName.tsx:1]: This is just extracting code from existing views

File: src/api/hooks/useFetchPerson.ts
Comment (denektenina) [inline src/api/hooks/useFetchPerson.ts:1]: This is just extracting code from existing views

File: src/pages/projects/form/ProjectFormDialog.tsx [hunk 1]
```diff
@@ -97,11 +97,11 @@ export const ProjectFormDialog = ({
 
   return (
     <Dialog
-      maxWidth="md"
+      maxWidth="lg"
```
Comment (denektenina) [inline src/pages/projects/form/ProjectFormDialog.tsx:100]: We need more space in the view to fit the big affiliation-boxes

File: src/pages/projects/form/ProjectFormDialog.tsx [hunk 2]
```diff
@@ -97,11 +97,11 @@ export const ProjectFormDialog = ({
...
       fullWidth
       onClose={toggleShowConfirmCloseDialog}
       open={open}
-      PaperProps={{ sx: { bgcolor: 'info.light' } }}
+      PaperProps={{ sx: { bgcolor: 'secondary.main' } }}
```
Comment (denektenina) [inline src/pages/projects/form/ProjectFormDialog.tsx:104]: New background color in the box

File: src/pages/projects/form/ProjectFormDialog.tsx [hunk 3]
```diff
@@ -130,10 +130,13 @@ export const ProjectFormDialog = ({
...
-                contributors: values.contributors.map(() => ({
-                  type: true,
+                contributors: values.contributors.map((contributor) => ({
                   identity: { id: true },
-                  affiliation: { id: true },
+                  roles: contributor.roles.map(() => ({
+                    affiliation: {
+                      id: true,
+                    },
+                  })),
```
Comment (denektenina) [inline src/pages/projects/form/ProjectFormDialog.tsx:139]: New structure for the touched-object to fit the new API backend. This is used when detecting if a user has visited a field, and thus if an error message should be displayed.

File: src/pages/projects/form/ProjectContributors.tsx [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+      <Typography variant="h3" gutterBottom sx={{ marginTop: '2rem', marginBottom: '1rem' }}>
+        {t('project.project_participants')}
+      </Typography>
+      {suggestedProjectManager && (
+        <Typography sx={{ marginBottom: '1rem' }}>
+          {t('project.project_manager_from_nfr', { name: suggestedProjectManager })}
+        </Typography>
+      )}
+      <FieldArray name={ProjectFieldName.Contributors}>
+        {({ name, remove, push }: FieldArrayRenderProps) => (
```
Comment (denektenina) [inline src/pages/projects/form/ProjectContributors.tsx:49]: This is from ProjectFormPanel1 with small changes

File: src/pages/registration/description_tab/projects_field/ProjectContributorRow.tsx [hunk 5]
```diff
@@ -1,249 +0,0 @@
...
-                    ? contributorErrors?.identity?.id
-                    : ''
-                }
-                isLoading={personQuery.isFetching}
-                showSearchIcon={!field.value.id}
-              />
-            )}
-          />
-        )}
-      </Field>
```
Comment (denektenina) [inline src/pages/registration/description_tab/projects_field/ProjectContributorRow.tsx:186]: Moved to `ContributorRow`

File: src/pages/projects/form/ContributorRow.tsx [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+                      : ''
+                  }
+                  isLoading={personQuery.isFetching}
+                  showSearchIcon={!field.value.id}
+                />
+              )}
+            />
+          )}
+        </Field>
+      </TableCell>
```
Comment (denektenina) [inline src/pages/projects/form/ContributorRow.tsx:159]: From the now deleted `ProjectContibutorRow`

File: src/pages/projects/form/ContributorRow.tsx [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+                unitUri={role.affiliation.id}
+                authorName={getFullName(contributor.identity.firstName, contributor.identity.lastName)}
+                contributorRoles={contributor.roles}
+                baseFieldName={baseFieldRoles}
+                sx={{ width: '100%', marginBottom: '0.5rem' }}
+                removeAffiliation={
+                  contributor.roles.length === 1 || role.type === 'ProjectManager'
+                    ? undefined
+                    : () => removeAffiliation(role.affiliation.id)
+                }
```
Comment (denektenina) [inline src/pages/projects/form/ContributorRow.tsx:176]: Each of the affiliation-boxes

File: src/pages/projects/form/ContributorRow.tsx [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+            </Typography>
+          )}
+        </>
+      </TableCell>
+      <TableCell sx={{ width: '3rem', textAlign: 'center', paddingTop: '1rem' }}>
+        <DeleteIconButton
+          data-testid={dataTestId.registrationWizard.description.projectForm.removeContributorButton}
+          onClick={() => setShowConfirmRemoveContributor(true)}
+          tooltip={t('project.form.remove_participant')}
+          disabled={!removeContributor}
```
Comment (elkebab) [inline src/pages/projects/form/ContributorRow.tsx:200]: Ser ut som denne alltid er disabled for det første elementet, mens den skulle vært disabled om det bare er ett element?
bilde

File: src/pages/projects/form/ProjectFormPanel1.tsx [hunk 9]
```diff
@@ -126,53 +124,8 @@ export const ProjectFormPanel1 = ({ currentProject, suggestedProjectManager }: P
...
-                {t('common.add')}
-              </Button>
-            </>
-          )}
-        </FieldArray>
-
-        <ProjectFundingsField currentFundings={values.funding} />
-      </Box>
+      <ProjectContributors currentProject={currentProject} suggestedProjectManager={suggestedProjectManager} />
+      <ProjectFundingsField currentFundings={values.funding} />
```
Comment (elkebab) [inline src/pages/projects/form/ProjectFormPanel1.tsx:128]: tror du noe av dette vil kunne være gjenbrukbart for project wizard også?

File: src/pages/projects/form/ProjectContributors.tsx
Comment (elkebab) [inline src/pages/projects/form/ProjectContributors.tsx:1]: I wizard står jeg oppført med en rolle og 2 tilknytninger, mens på landing page kommer den ene tilknytningen opp under en annen rolle. Usikker på hvor mye vi skal legge i det, men visning på landing page får sikkert samme utfordring etter vi er over på wizard også?

bilde
bilde

Comment (elkebab) [reply src/pages/projects/form/ProjectContributors.tsx:1]: Her er ett eksempel fra cristin, der en person har to tilknytninger (siste deltageren): https://app.cristin.no/projects/show.jsf?id=464212
Men ser nå at den fungerer hos oss i dev, ved at den bare dukker opp en gang: https://dev.nva.sikt.no/projects?id=https%3A%2F%2Fapi.dev.nva.aws.unit.no%2Fcristin%2Fproject%2F464212

Så det er kanskje noe som skjer bare om det er prosjektleder?

File: src/pages/projects/form/ContributorRow.tsx (hunk 8 above)
Comment (denektenina) [reply src/pages/projects/form/ContributorRow.tsx:200]: Nei, logikken er kodet slik at den også er disablet hvis det er rollen til projectmanager.

Comment (denektenina) [reply src/pages/projects/form/ContributorRow.tsx:200]: Mulig NTNU kunne rykket opp til å bli project manager-rolle, men det er litt skummelt å gjøre så mye logikk som ikke vises i frontend syns jeg

File: src/pages/projects/form/ProjectFormPanel1.tsx (hunk 9 above)
Comment (denektenina) [reply src/pages/projects/form/ProjectFormPanel1.tsx:128]: Ja heldigvis!

File: src/pages/projects/form/ProjectContributors.tsx
Comment (denektenina) [reply src/pages/projects/form/ProjectContributors.tsx:1]: Sånn som jeg forsto ønsket fra Mayli og JE, skulle vi nå liste ut alle roller som tilknytninger i denne versjonen. I neste versjon i wizard vil rolle-feltet liste ut alle roller man har om man f.eks har både prosjektleder og prosjektdeltaker. 

image

File: src/pages/projects/form/ContributorRow.tsx (hunk 8 above)
Comment (elkebab) [reply src/pages/projects/form/ContributorRow.tsx:200]: Ah ja, for nå ligger jeg inne som prosjektleder i Sikt, mens jeg er prosjektdeltager i NTNU?
Det tror jeg egentlig er feil da, og at i dette tilfelle er jeg bare prosjektleder (ikke både leder og deltager), men jeg er det på vegne av to ulike enheter.
Det beste hadde kanskje vært å bruke samme rolle som for den første tilknytningen også for alle andre tilknytninger
Men om vi har en plan for å løse dette i wizarden er det greit for meg

File: src/pages/projects/form/ProjectContributors.tsx
Comment (elkebab) [reply src/pages/projects/form/ProjectContributors.tsx:1]: Det som var rart her for meg er at jeg vises under to roller, men i formen har jeg bare en rolle (Prosjektleder). Som i den linken til dagens løsning i cristin har alle eksemplene jeg har funnet hvor folk har hatt flere tilknytninger, alltid vært koblet til en rolle, og dermed vil personene bare vises en gang på landing page. Vi åpner nå for at en person kan være både leder og deltager. Men skjønner nå at dette egentlig er samme sak som https://github.com/BIBSYSDEV/NVA-Frontend/pull/6048#discussion_r1721292782, så greit nok for meg

**Review by elkebab — APPROVED** [review]
Ser veldig bra ut det!

For mye kode til at jeg har ork og tid til å se på det, så har bare sett på funksjonalitet. Liten NB ved at vi nå tillater samme person å ha ulike roller, men ellers good!

File: src/pages/projects/form/ContributorRow.tsx (hunk 8 above)
Comment (denektenina) [reply src/pages/projects/form/ContributorRow.tsx:200]: Jeg spurte ihvertfall JE om det var greit at rolle nr to alltid var "Prosjektdeltaker" og han sa ja på det og at majoriteten alltid skulle være prosjektdeltaker.

**Pull request merged**